### ISSUE TITLE
[Client emitter] Generate a metadata file while emitting

### ISSUE THREAD

**Issue by lmazuel** [Azure/typespec-azure#1758]
We want to save some metadata in a file along with the emitted code, that can be used for some automation purposes. For instance, @weshaggard would want to use it to build a mapping of package version with support api-version

# Proposal

Create a file `metadata.json` that contains:
```json
{
  "api-version": "2015-06-01"
}
```

This file is not generated by default, and is enabled if the flag `emit-metadata: true` is passed in the tspconfig.yaml.

Each emitter will need to do that code, as TCGC should not be responsible to write files on disk. We could have though TCGC building the dict for us, with a `getMetadataInfo()` that emitters just have to put on disk. This way, this is consistent cross-language.

**Comment by msyyc** [comment]
I think the proposal makes sense. About the filename, since SDK users usually don't need care about it, we could mark it with `_` to declare it is inner file so maybe `_metadata.json` is better.

**Comment by tadelesh** [comment]
though tcgc should be a lib depended by language's emitter, it has the ability to emit the metadata file if needed.

**Comment by lmazuel** [comment]
@srnagar is suggesting we're merging that with the current `apiview_properties.json` file, and I feel it makes sense to avoid having too much files

**Comment by tadelesh** [comment]
> [@srnagar](https://github.com/srnagar) is suggesting we're merging that with the current `apiview_properties.json` file, and I feel it makes sense to avoid having too much files

agree. emitter could use `client.apiVersions[-1]` to get the api version. @iscai-msft do you think we need to add more explicit property in `SdkClientType` to indicate the api version client generated from?

**Comment by iscai-msft** [comment]
we have `getDefaultApiVersion` right now, I feel that's the best way to do it. We could add a `defaultApiVersion` property, but I think we're good for now

**Comment by tadelesh** [comment]
`getDefaultApiVersion` need the raw typespec namespace. i don't think it is a good way to get default.

**Comment by iscai-msft** [comment]
Hmm that's a very good point, yes we can go with `client.apiVersions[-1]` for now / add another property, but I think we can go without adding the property for now

**Comment by tadelesh** [comment]
TCGC will export `apiVersion` in `SdkPacakge`.

**Comment by MaryGao** [comment]
I reopened this issue to discuss some decisions cross languages and @iscai-msft  let me know if you have some concern!

**Comment by MaryGao** [comment]
@weshaggard I need your input regarding this work. Two things we'd like to confirm with you:

- The metadata file would be JSON format and each language emitter would include at least api-version information which is a JSON string type, but except that emitters may put some other information for their purpose. Is it okay for you?
- Some language emitters have concerns to include this in public released SDK, so we may prefer to only include this file in SDK repo's main branch. Does this work for to collect the data?

**Comment by weshaggard** [comment]
- json format is fine and I'm also good with it containing other information if that is useful for other purposes. 
- What is the concern if this metadata file ships in the SDK packages? It might a helpful clue for folks. However, for my purposes I just need it to live in the repo next to the package so that we can index the information and show what api-versions a given SDK supports.

**Comment by lmazuel** [comment]
> * json format is fine and I'm also good with it containing other information if that is useful for other purposes.
> * What is the concern if this metadata file ships in the SDK packages? It might a helpful clue for folks. However, for my purposes I just need it to live in the repo next to the package so that we can index the information and show what api-versions a given SDK supports.

Python has two deliverables: the sdist (source distribution) and the wheel (runtime distribution). I'm fine putting it in the sdist, but I'd like to avoid make the wheel "dirty" with files that are not runtime. I don't see the benefits compare to violating a tacit Python guidelline.

**Comment by MaryGao** [comment]
> What is the concern if this metadata file ships in the SDK packages? It might a helpful clue for folks. However, for my purposes I just need it to live in the repo next to the package so that we can index the information and show what api-versions a given SDK supports.

For JS the distribution would be public for customers and here is an example for [arm compute](https://www.npmjs.com/package/@azure/arm-compute?activeTab=code). Since this is an internal usage and we don't want to pollute our distribution. I know Go has the similar concern /cc @tadelesh.

**Comment by tadelesh** [comment]
yes. go does not want to package such metadata that is not useful for customer.

**Comment by MaryGao** [comment]
Let's conclude with following decisions:
- The filename would be `_metadata.json` and it would live in the SDK repo under the root folder of the package;
- The file would be JSON format and at least the `api-version` property. Currently we only consider the signle api-version case

```
{
  "api-version": "2015-06-01"
  // language emitters could add more data for their purpose
}
```

**Comment by weidongxu-microsoft** [comment]
> Let's conclude with following decisions:
> 
> * The filename would be `_metadata.json` and it would live in the SDK repo under the root folder of the package;
> * The file would be JSON format and at least the `api-version` property. And for api-version we only consider the signle api-version for now. Language emitters could add other data for their purpose.
> 
> ```
> {
>   "api-version": "2015-06-01"
>   // language emitters could add more data for their purpose
> }
> ```

Loop @srnagar for opinion.
1. the file is not included in release
2. it be `_metadata.json` in module folder

As it is not included in released package, the only guaranteed "correct" value would be the file in the release tag of that specific package.
The file/value on main branch could be unreleased (service generate/merge the SDK, but yet to release).

---

Unrelated.

Java also include the `api-version` in pom description, [example](https://github.com/Azure/azure-sdk-for-java/blob/0a0edd32ba518c53aae6d16b58c408c552753e54/sdk/servicenetworking/azure-resourcemanager-servicenetworking/pom.xml#L21)
Previously in Swagger, we put the [tag for reference](https://central.sonatype.com/artifact/com.azure.resourcemanager/azure-resourcemanager-mobilenetwork/1.2.0).
![Image](https://github.com/user-attachments/assets/a61515d0-6e53-4d10-ac7e-19cd48bb9970)

**Comment by tadelesh** [comment]
cc: @JiaqiZhang-Dev

**Comment by srnagar** [comment]
> > Let's conclude with following decisions:
> > 
> > * The filename would be `_metadata.json` and it would live in the SDK repo under the root folder of the package;
> > * The file would be JSON format and at least the `api-version` property. And for api-version we only consider the signle api-version for now. Language emitters could add other data for their purpose.
> > 
> > ```
> > {
> >   "api-version": "2015-06-01"
> >   // language emitters could add more data for their purpose
> > }
> > ```
> 
> Loop [@srnagar](https://github.com/srnagar) for opinion.
> 
> 1. the file is not included in released package
> 2. it be `_metadata.json` in module/package folder root
> 
> As it is not included in released package, the only guaranteed "correct" value would be the file in the release tag of that specific package. The file/value on main branch could be unreleased (service generate/merge the SDK, but yet to release).
> 
> Unrelated.
> 
> Java also include the `api-version` in pom description, [example](https://github.com/Azure/azure-sdk-for-java/blob/0a0edd32ba518c53aae6d16b58c408c552753e54/sdk/servicenetworking/azure-resourcemanager-servicenetworking/pom.xml#L21) Previously in Swagger, we put the [tag for reference](https://central.sonatype.com/artifact/com.azure.resourcemanager/azure-resourcemanager-mobilenetwork/1.2.0) (as readme tag be a collection of API and versions). ![Image](https://github.com/user-attachments/assets/a61515d0-6e53-4d10-ac7e-19cd48bb9970)

For Java, we should include the metadata file in the jar file. It should be in the resources directory similar to the apiview_properties.json file. We should also consider combining the metadata and apiview propertis json file.

**Comment by weidongxu-microsoft** [comment]
As an example on Srikanta's comment, Java would output it at `src/main/resources/META-INF` of the module/package root. E.g.,
https://github.com/Azure/azure-sdk-for-java/tree/main/sdk/devcenter/azure-developer-devcenter/src/main/resources/META-INF

If use a separate file, we are also considering put a prefix on the filename e.g. `azure-developer-devcenter_metadata.json`

**Comment by weshaggard** [comment]
The exact location in the repo can be configurable but I would like to try and avoid different file names if possible (although package name might be OK). For my consumption I want to create a link template similar to these others (https://github.com/Azure/azure-sdk/edit/main/_includes/releases/variables/java.md) and be able to find the metadata file based on the source location, package name (and groupId for java) and version as the only parameters. If I can build a predictable path with that information my indexing will work.

**Comment by weshaggard** [comment]
Reopening as I suspect this got accidently closed as it was mentioned in a PR.

**Comment by anuchandy** [comment]
Regarding metadata.json vs {package-prefix-}metadata.json for Java:

When running a Java app in a hosted environment, it is common to combine multiple jar files (az-storage.jar, az-cosmos.jar) into one shaded jar file (contoso-app.jar). During this shading process, resources/META-INF files from different jars are included in the final shaded jar. If two jars contain a resource with the same name (e.g., metadata.json), the last processed one will overwrite earlier ones. Using unique package-prefixed names for metadata can prevent this. If meta includes debug info like compiler or TCGC versions, which may be different for different individual jars, it may be lost on overwrite. Customers can change default overwrite to append (though this may produce invalid final JSON). Just thinking loud.

**Comment by weshaggard** [comment]
I can get behind using the artifact and/or group id as part of the file name.

**Comment by JoshLove-msft** [comment]
@weshaggard is it sufficient that the file exists in the repo? Presumably you can get to the specific version of the file for a release from the corresponding release tag.

**Comment by weshaggard** [comment]
Yes we can use release tags for getting the specific version of the metadata file. We just need to have a template that takes the package name and version and can produce a path to a file to get the metadata.

**Comment by MaryGao** [comment]
>  I would like to try and avoid different file names if possible (although package name might be OK)
@weshaggard Can I understand this with followings?
- It's better to align the filename cross languages like `metadta.json`;
- But it's okay if some languages would like to add package name prefix that;
- Also it's okay if some languages don't add any prefix

**Comment by weshaggard** [comment]
My preference would be to have all them files named the same but there are reason like Java stated where that will be problematic, so I'm good with that. As long as we only need to the package name and version to find the path of the file as I need to build a template for each language in order to find the associated metadata file for each package version.

**Comment by MaryGao** [comment]
Updated the MoM for today's discussion:

- Azure always generate the file (repo or not)
- Unbranded should not support the feature
- TCGC should provide the content to put in the file (emitter don't have to do it themselves)
- If we have Azure crosslanguageid, it should go in that file. If you support the file in unbranded, this should not include crosslanguageid
- No `emit-metadata`  option for now and use the flavor as the `emit-metadata` flag
- The filename would match the pattern of `*metadata.json`, for example Python would use _metadata.json but Java would be {package-prefix-}metadata.json
- Regarding the path location Java would output it at src/main/resources/META-INF of the module/package root but for other languagues the file location would be in the package root
- No need to consider multiple api-versions
- The JSON field names should be camel case
- If we would emit language emitter version, we could output as `emitterVersion`

**Comment by weshaggard** [comment]
Sounds like a good summary. 

For the json fields name I would suggest using lowercase and dashes like in the proposal so "api-version" or "emitter-version" instead of camel casing.

**Comment by MaryGao** [comment]
> DO use camel case for all JSON field names.

@weshaggard we discussed this and we'd like to follow the REST API guidelines to suggest using [camel casing for JSON field names](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#json-field-name-casing). That's why we propose as "apiVersion" or "emitterVersion".

**Comment by weshaggard** [comment]
Thanks for the context and I'm find with that as long as we are consistent.

**Comment by weidongxu-microsoft** [comment]
On
> If we have Azure crosslanguageid, it should go in that file. If you support the file in unbranded, this should not include crosslanguageid

We will make is consistent to use camelCase `crossLanguageDefinitions` on the JSON property name.
E.g. https://github.com/Azure/azure-sdk-for-java/blob/c85198c1fdd7329f2acb16da7fa367097230ef40/sdk/standbypool/azure-resourcemanager-standbypool/src/main/resources/META-INF/azure-resourcemanager-standbypool_metadata.json
Let me know if there is concern about the casing of property name.

The existing "apiview_properties.json" is kept for now, till APIView done the switch to "metadata.json".
E,g, https://github.com/Azure/azure-sdk-for-java/blob/c85198c1fdd7329f2acb16da7fa367097230ef40/sdk/standbypool/azure-resourcemanager-standbypool/src/main/resources/META-INF/azure-resourcemanager-standbypool_apiview_properties.json

**Comment by tadelesh** [comment]
close since all languages get consensus on the impl.

**Comment by weidongxu-microsoft** [comment]
Java completed and released.

metadata.json in main https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/standbypool/azure-resourcemanager-standbypool/src/main/resources/META-INF/azure-resourcemanager-standbypool_metadata.json
